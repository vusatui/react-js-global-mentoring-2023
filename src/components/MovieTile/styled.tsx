"use client";

import styled from "styled-components";

export const StyledMovieCard = styled.div.attrs( { "data-testid": "movie-card" })`
  position: relative;
  width: 323px;
  cursor: pointer;
`;

export const StyledCardTop = styled.div`
  position: relative;
  margin-bottom: 20px;
`;

export const StyledImage = styled.img`
  width: 100%;
`;

export const StyledCardBottom = styled.div`

`;

export const StyledCardBottomTop = styled.div`
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 8px;
`;


export const StyledCardTitle = styled.div`
  font-family: 'Montserrat', serif;
  font-style: normal;
  font-weight: 500;
  font-size: 18px;
  line-height: 22px;
  color: #FFFFFF;
  opacity: 0.7;
`;

export const StyledCardYear = styled.div`
  border: 1px solid #555555;
  border-radius: 4px;
  padding: 4px 8px;
  font-family: 'Montserrat', serif;
  font-style: normal;
  font-weight: 500;
  font-size: 14px;
  line-height: 17px;
  text-align: center;
  color: #FFFFFF;
  mix-blend-mode: normal;
  opacity: 0.7;
`;

export const StyledCardGenres = styled.div`
  font-family: 'Montserrat', serif;
  font-style: normal;
  font-weight: 500;
  font-size: 14px;
  line-height: 17px;
  color: #FFFFFF;
  mix-blend-mode: normal;
  opacity: 0.5;
`;