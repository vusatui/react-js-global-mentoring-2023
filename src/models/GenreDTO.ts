export interface GenreDTO {
    name: string
}