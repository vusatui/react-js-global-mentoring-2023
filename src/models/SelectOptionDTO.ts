export interface SelectOptionDTO<T> {
    id: string
    label: string
    value: T
}