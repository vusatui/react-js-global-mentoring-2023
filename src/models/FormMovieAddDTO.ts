import {GenreDTO} from "./GenreDTO";

export interface FormMovieAddDTO {
    title: string
    movieUrl: string
    genre: GenreDTO[]
    overview: string
    runtime: string
    rating: string
    releaseDate: string
}