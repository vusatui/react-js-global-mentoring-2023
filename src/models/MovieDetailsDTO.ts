import {GenreDTO} from "./GenreDTO";

export interface MovieDetailsDTO {
    imageUrl: string
    title: string
    year: string
    genres: GenreDTO[]
    duration: string
    rating: number
    description: string
}