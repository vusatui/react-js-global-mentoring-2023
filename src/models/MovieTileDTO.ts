import {GenreDTO} from "./GenreDTO";

export interface MovieTileDTO {
    imageUrl: string
    title: string
    year: string
    genres: GenreDTO[]
}