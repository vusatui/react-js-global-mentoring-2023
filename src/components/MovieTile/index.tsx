import {useMemo} from "react";

import {MovieTileDTO} from "../../models/MovieTileDTO";

import {
    StyledCardBottom,
    StyledCardBottomTop, StyledCardGenres,
    StyledCardTitle,
    StyledCardTop, StyledCardYear,
    StyledImage,
    StyledMovieCard
} from "./styled";

interface MovieTileProps {
    className?: string
    movieTile: MovieTileDTO
    onClick: (movieTile: MovieTileDTO) => void
}

const GENRE_SEPARATOR = ", ";

const MovieTile = ({
    className,
    movieTile,
    onClick
}: MovieTileProps) => {
    const handleClick = () => onClick(movieTile);

    const genres = useMemo(() => movieTile.genres.map(genre => genre.name)
        .join(GENRE_SEPARATOR), [movieTile])

    return (
        <StyledMovieCard
            className={className}
            onClick={handleClick}
        >
            <StyledCardTop>
                <StyledImage src={movieTile.imageUrl} />
            </StyledCardTop>
            <StyledCardBottom>
                <StyledCardBottomTop>
                    <StyledCardTitle>
                        {movieTile.title}
                    </StyledCardTitle>
                    <StyledCardYear>
                        {movieTile.year}
                    </StyledCardYear>
                </StyledCardBottomTop>
                <StyledCardGenres>
                    {genres}
                </StyledCardGenres>
            </StyledCardBottom>
        </StyledMovieCard>
    );
}

export default MovieTile;