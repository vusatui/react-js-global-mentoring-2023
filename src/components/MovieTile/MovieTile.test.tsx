import MovieTile from "../MovieTile";
import {GenreDTO} from "../../models/GenreDTO";
import {MovieTileDTO} from "../../models/MovieTileDTO";
import {render, screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

const genres: GenreDTO[] = [
    { name: "Drama" },
    { name: "Biography" },
    { name: "Music" },
];

const movieTileDTO: MovieTileDTO = {
    imageUrl: "https://upload.wikimedia.org/wikipedia/en/2/2e/Bohemian_Rhapsody_poster.png",
    title: "Bohemian Rhapsody",
    year: "2003",
    genres,
};

const mockHandleClick = jest.fn();

const Component = () => (<MovieTile movieTile={movieTileDTO} onClick={mockHandleClick} />);

describe("MovieTile", () => {

    it("Should renders correctly", () => {
        const { asFragment } = render(<Component/>);
        expect(asFragment()).toMatchSnapshot();
    });

    it("Should on click receive correct arg", () => {
        render(<Component/>);
        userEvent.click(screen.getByTestId("movie-card"));
        expect(mockHandleClick).toBeCalledWith(movieTileDTO);
    });

});