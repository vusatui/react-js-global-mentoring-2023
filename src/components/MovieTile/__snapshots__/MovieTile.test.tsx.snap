// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieTile Should renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-beqWaB vaRaK"
    data-testid="movie-card"
  >
    <div
      class="sc-gueYoa iwxqtk"
    >
      <img
        class="sc-dmqHEX XlOMr"
        src="https://upload.wikimedia.org/wikipedia/en/2/2e/Bohemian_Rhapsody_poster.png"
      />
    </div>
    <div
      class="sc-hLseeU jmgCIT"
    >
      <div
        class="sc-eDDNvR gUHsQd"
      >
        <div
          class="sc-jTrPJq PPPYC"
        >
          Bohemian Rhapsody
        </div>
        <div
          class="sc-gLDzan eVTOHm"
        >
          2003
        </div>
      </div>
      <div
        class="sc-iAEyYk fjgTJV"
      >
        Drama, Biography, Music
      </div>
    </div>
  </div>
</DocumentFragment>
`;
